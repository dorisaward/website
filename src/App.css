#root {
  flex: 1;
  height: 100%;
  text-align: center;
  background-color: #cce6ff;
}
