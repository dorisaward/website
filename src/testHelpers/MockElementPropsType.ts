import { ReactNode } from 'react'

export type MockElementPropsType<P = unknown> = P & { children?: ReactNode | undefined }
