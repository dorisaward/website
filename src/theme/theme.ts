import { DefaultTheme } from 'styled-components'

export enum theme {
    light = 'LIGHT_THEME',
    dark = 'DARK_THEME',
}

const commonTheme: Pick<DefaultTheme, 'padding'
    | 'switchOnColour'
    | 'switchOffColour'
    | 'shadowColour'
> = {
    padding: '5px',
    switchOnColour: 'green',
    switchOffColour: 'grey',
    shadowColour: 'chocolate',
}

export const lightTheme: DefaultTheme = {
    ...commonTheme,
    backgroundColour: '#CCE6FF',
    secondaryColour: '#FFCCE6',
    tertiaryColour: '#E6FFCC',
    fontColour: 'black'
}

export const darkTheme: DefaultTheme = {
    ...commonTheme,
    backgroundColour: '#214269',
    secondaryColour: '#692142',
    tertiaryColour: '#426921',
    fontColour: 'white'
}
