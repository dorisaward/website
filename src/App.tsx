import { useEffect, useState } from 'react'
import styled, { ThemeProvider } from 'styled-components'
import { darkTheme, lightTheme } from './theme/theme'
import { Home } from './pages/homepage/Home'
import { Switch } from './assets/Switch'

const Container = styled.div`
  position: absolute;
  top: 0;
  bottom: 0;
  width: 100%;
  text-align: center;
  background-color: ${({ theme }) => theme.backgroundColour};
  color: ${({ theme }) => theme.fontColour};

  font-family: Inter, system-ui, Avenir, Helvetica, Arial, sans-serif;
  line-height: 1.5;
  font-weight: 400;
  font-synthesis: none;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
`

const displayAnimatedFaviconIfBrowserNotChrome = () => {
  const userAgentString = navigator.userAgent
  const isChrome = userAgentString.indexOf("Chrome") > -1
  const faviconElement: any = document.querySelector('link[rel=icon]')
  if (!isChrome && !!faviconElement) {
    faviconElement.href = '/src/assets/daw.gif'
    faviconElement.type = 'image/gif'
  }
}

enum theme {
  light = 'LIGHT_THEME',
  dark = 'DARK_THEME',
}

function App() {
  const [selectedTheme, setSelectedTheme] = useState<theme>(theme.light)
  useEffect(displayAnimatedFaviconIfBrowserNotChrome)

  const handleThemePress = () => setSelectedTheme(prevTheme => prevTheme === theme.light ? theme.dark : theme.light)

  return (
    <ThemeProvider theme={selectedTheme === theme.light ? lightTheme : darkTheme}>
      <Container>
        <Home />
        <Switch checked={selectedTheme === theme.dark} onSwitch={handleThemePress} label={{ left: '🌘', right: '☀️' }} />
      </Container>
    </ThemeProvider>
  )
}

export default App
