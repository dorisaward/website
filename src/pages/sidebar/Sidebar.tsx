import styled from 'styled-components'
import { use, useState } from 'react'
import { Outlet } from 'react-router-dom'
import { Switch } from '../../assets/Switch.tsx'
import { theme } from '../../theme/theme.ts'
import { Hamburger } from '../../assets/Hamburger.tsx'
import { LanguagesContext } from '../../languages/LanguagesContext.ts'
import { languages } from '../../languages/languages.ts'
import { Navigation } from '../../assets/Navigation.tsx'
import { MOBILE_WIDTH } from '../../assets/constants.ts'

const SWITCH_WIDTH = '120px'
const HAMBURGER_SIZE = 24

const AppContainer = styled.div`
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    text-align: center;
    background-color: ${({ theme }) => theme.backgroundColour};
    color: ${({ theme }) => theme.fontColour};
    
    font-family: Inter, system-ui, Avenir, Helvetica, Arial, sans-serif;
    line-height: 1.5;
    font-weight: 400;
    font-synthesis: none;
    text-rendering: optimizeLegibility;
    -webkit-font-smoothing: antialiased;
    -moz-osx-font-smoothing: grayscale;
`

const Container = styled.div<{ isVisible: boolean }>`
    background: ${({ theme }) => theme.secondaryColour };
    display: flex;
    flex-wrap: wrap;
    position: fixed;
    width: 100vw; // mobile-only
    top: 0;
    padding-top: ${HAMBURGER_SIZE + 10}px;
    margin-left: ${({ isVisible }) => (isVisible ? 0 : -100) + '%'};
    transition: margin-left 0.1s;
    @media only screen and (min-width: ${MOBILE_WIDTH}px) {
        width: ${SWITCH_WIDTH};
        margin-left: ${({ isVisible }) => (isVisible ? 0 : '-' + SWITCH_WIDTH)};
    }
`
const HamburgerContainer = styled.div<{ isVisible: boolean }>`
    background: white;
    padding: ${({theme}) => theme.padding};
    width: ${HAMBURGER_SIZE}px;
    height: ${HAMBURGER_SIZE}px;
    position: absolute;
    top: 0;
    margin-left: ${({ isVisible }) => (isVisible ? 0 : 100) + '%'};
    @media only screen and (min-width: ${MOBILE_WIDTH}px) {
        margin-left: ${SWITCH_WIDTH};
    }
`
const ItemContainer = styled.div`
    flex-grow: 1;
    flex-basis: ${SWITCH_WIDTH};
`

export type SidebarProps = {
    theme: {
        selectedTheme: theme
        handleThemePress: () => void
    }
}

export const Sidebar = ({
    theme: { selectedTheme, handleThemePress }
}: SidebarProps) => {
    const [isVisible, setIsVisible] = useState(false)
    const { language, handleLangPress } = use(LanguagesContext)
    const toggleVisibility = () => setIsVisible(prevVisible => !prevVisible)

    const RenderHamburger = () => (
        <HamburgerContainer onClick={toggleVisibility} isVisible={isVisible}>
            <Hamburger
                strokeWidth={4}
                height={HAMBURGER_SIZE}
                width={HAMBURGER_SIZE}
            />
        </HamburgerContainer>
    )

    return (
        <AppContainer>
            <Container isVisible={isVisible}>
                <RenderHamburger />
                <ItemContainer>
                    <Navigation/>
                </ItemContainer>
                <ItemContainer>
                    <Switch
                        checked={selectedTheme === theme.dark}
                        onSwitch={handleThemePress}
                        label={{left: '🌘', right: '☀️'}}
                    />
                    <Switch
                        checked={language === languages.en}
                        onSwitch={handleLangPress}
                        label={{left: 'UK', right: 'TH'}}
                    />
                </ItemContainer>
            </Container>
            <Outlet/>
        </AppContainer>
    )
}