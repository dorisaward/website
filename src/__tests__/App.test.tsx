import { fireEvent, render } from '@testing-library/react'
import App from '../App.tsx'
import { MockElementPropsType } from '../testHelpers/MockElementPropsType.ts'
import { ReactNode } from 'react'
import { SidebarProps } from '../pages/sidebar/Sidebar.tsx'
import { theme } from '../theme/theme.ts'

jest.mock('../pages/homepage/Home.tsx', () => ({
    Home: ({ children, ...restProps }: MockElementPropsType) => <div {...restProps}>{children}</div>,
}))
const CHANGE_THEME_BUTTON_TEXT = 'CHANGE_THEME_BUTTON'
jest.mock('../pages/sidebar/Sidebar.tsx', () => ({
    Sidebar: ({ theme }: MockElementPropsType<SidebarProps>) => (
        <div>
            <button onClick={theme.handleThemePress}>{CHANGE_THEME_BUTTON_TEXT}</button>
            <p>{theme.selectedTheme}</p>
        </div>
    ),
}))
jest.mock('../pages/cv/Cv.tsx', () => ({
    Cv: ({ children, ...restProps }: MockElementPropsType) => <div {...restProps}>{children}</div>,
}))

jest.mock('react-router-dom', () => ({
    BrowserRouter: ({ children, ...restProps }: MockElementPropsType) => <div {...restProps}>{children}</div>,
    Routes: ({ children, ...restProps }: MockElementPropsType) => <div {...restProps}>{children}</div>,
    Route: ({ children, element, ...restProps }: MockElementPropsType<{element?: ReactNode}>) => (
        <div {...restProps}>{element}{children}</div>
    ),
}))

describe('App', () => {
    it('renders', () => {
        // Given
        const { baseElement } = render(<App/>)

        // Then
        expect(baseElement).toMatchSnapshot()
    })

    it('switches theme', () => {
        // Given
        const { getByText, queryByText } = render(<App/>)

        // When
        fireEvent.click(getByText(CHANGE_THEME_BUTTON_TEXT))

        // Then
        expect(queryByText(theme.light)).toBeFalsy()
        expect(getByText(theme.dark)).toBeTruthy()
    })

    it('switches theme back', () => {
        // Given
        const { getByText, queryByText } = render(<App/>)

        // When
        fireEvent.click(getByText(CHANGE_THEME_BUTTON_TEXT))
        fireEvent.click(getByText(CHANGE_THEME_BUTTON_TEXT))

        // Then
        expect(getByText(theme.light)).toBeTruthy()
        expect(queryByText(theme.dark)).toBeFalsy()
    })
})