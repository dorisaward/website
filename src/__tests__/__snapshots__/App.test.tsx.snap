// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders 1`] = `
<body>
  <div>
    <div>
      <div>
        <div
          path="/"
        >
          <div>
            <button>
              CHANGE_THEME_BUTTON
            </button>
            <p>
              LIGHT_THEME
            </p>
          </div>
          <div>
            <div />
          </div>
          <div
            path="cv"
          >
            <div />
          </div>
          <div
            path="*"
          >
            <p>
              error
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
